lib/compress/zstd_compress.c

```diff
@@ -2803,18 +2884,684 @@ static ZSTD_blockCompressor ZSTD_selectBlockCompressor(ZSTD_strategy strat, int
+static void ZSTD_ldm_limitedNextToUpdate(ZSTD_CCtx* cctx, const BYTE* anchor)
+{
+    U32 const current = (U32)(anchor - cctx->base);
+    if (current > cctx->nextToUpdate + 384) {
+        cctx->nextToUpdate =
+            current - MIN(192, current - cctx->nextToUpdate - 394);
+    }
+}
+
+FORCE_INLINE
+size_t ZSTD_compressBlock_ldm_generic(ZSTD_CCtx* cctx,
+                                      const void* src, size_t srcSize)
+{
+    ldmEntry_t* const hashTable = cctx->ldmState.hashTable;
+    BYTE* const bucketOffsets = cctx->ldmState.bucketOffsets;
+    U32  const hBits = LDM_HASH_LOG - LDM_BUCKET_SIZE_LOG;
+    seqStore_t* seqStorePtr = &(cctx->seqStore);
+    const BYTE* const base = cctx->base;
+    const BYTE* const istart = (const BYTE*)src;
+    const BYTE* ip = istart;
+    const BYTE* anchor = istart;
+    const U32   lowestIndex = cctx->dictLimit;
+    const BYTE* const lowest = base + lowestIndex;
+    const BYTE* const iend = istart + srcSize;
+    const BYTE* const ilimit = iend - LDM_MIN_MATCH_LENGTH;
+
+    U32* const repToConfirm = seqStorePtr->repToConfirm;
+    U32 savedRep[ZSTD_REP_NUM];
+    U64 rollingHash;
+    const BYTE* lastHashed;
+    ZSTD_blockCompressor blockCompressor =
+        ZSTD_selectBlockCompressor(cctx->appliedParams.cParams.strategy, 0);
+    size_t i, lastLiterals;
+
+    /* Save seqStorePtr->rep and copy repToConfirm */
+    for (i = 0; i < ZSTD_REP_NUM; i++)
+        savedRep[i] = repToConfirm[i] = seqStorePtr->rep[i];
+
+    rollingHash = ZSTD_ldm_getRollingHash(ip, LDM_MIN_MATCH_LENGTH);
+    lastHashed = ip;
+
+    ip += (ip==lowest);
+
+    /* Main Search Loop */
+    while (ip < ilimit) {   /* < instead of <=, because repcode check at (ip+1) */
+        size_t mLength;
+        U32 const current = (U32)(ip - base);
+        size_t forwardMatchLength = 0, backwardMatchLength = 0;
+        ldmEntry_t* bestEntry = NULL;
+
+        if (ip != istart) {
+            rollingHash = ZSTD_ldm_updateHash(rollingHash, lastHashed[0],
+                                              lastHashed[LDM_MIN_MATCH_LENGTH]);
+        }
+
+        lastHashed = ip;
+
+        /* Do not insert and do not look for a match */
+        if (ZSTD_ldm_getTag(rollingHash, hBits) != LDM_HASH_MASK) {
+           ip++;
+           continue;
+        }
+
+        /* Get the best entry and compute the match lengths */
+        {
+            ldmEntry_t* const bucket =
+                ZSTD_ldm_getBucket(hashTable,
+                                   ZSTD_ldm_getSmallHash(rollingHash, hBits));
+            ldmEntry_t* cur;
+            size_t bestMatchLength = 0;
+            U32 const checksum = ZSTD_ldm_getChecksum(rollingHash, hBits);
+
+            for (cur = bucket; cur < bucket + LDM_BUCKET_SIZE; ++cur) {
+                const BYTE*const  pMatch = cur->offset + base;
+                if (cur->checksum != checksum || cur->offset <= lowestIndex) {
+                    continue;
+                } else {
```

Comment:
I'd delete the `else` to reduce the indentation a bit, and just declare the match lengths at the top of the loop.